### ISSUE TITLE
segfault for nastruct with dummy traj

### ISSUE THREAD

**Issue by hainm** [Amber-MD/cpptraj#168]
```
parm tz2.truncoct.parm7
trajin tz2.truncoct.nc 1 1
autoimage
nastruct
```

tail of out
```
ACTION SETUP FOR PARM 'tz2.truncoct.parm7' (2 actions):
  0: [autoimage]
        Original box is truncated octahedron, turning on 'familiar'.
        Anchor molecule is 1
        1869 molecules are mobile.
  1: [nastruct]
        Set up 0 bases.
Segmentation fault
```

**Comment by swails** [comment]
Do you have a stack trace for the segfault?

**Comment by hainm** [comment]
@swails I don't know what you mean.

just `../../cpptraj/bin/cpptraj.OMP -i tz2.in`

I think cpptraj try to deallocate of never-allocated memory?

**Comment by hainm** [comment]
so is there any way (key word) I can add to monitor stack trace in cpptraj?

**Comment by swails** [comment]
What I do when I get a segfault is to compile with the "-g" flag turned on (which adds debug symbols), then issue the command `ulimit -c unlimited`.  This means that a core dump will be written no matter how large it is (I think default size is 0, so that a `core.pid` file is never written).  You can then load this into gdb and type `bt` to get a stack trace (which will tell you the exact source line where the segfault occurred with the whole call stack).

If you don't want to deal with the core file (which is really the best way to debug *parallel* segfaults), you can run cpptraj in gdb and just use `run <arguments>`.  The debugger will stop at the segfault, and you can type `bt` to get the stack trace.

This removes the guessing, usually, and usually lets you debug issues yourself.  If you still can't figure it out (as has happened to me a number of times), I post with the full stack trace, which helps whoever *does* figure it out quite a bit.

**Comment by swails** [comment]
First thing to try -- does it segfault with `cpptraj`, or just `cpptraj.OMP`?  If it's the latter, any debugging that looks at a process-by-process runtime will most likely be needed to debug.  In which case, you're left with analyzing the core file or using valgrind with the `--trace-children` and `--log-file=valgrind.%p.log` (so you get separate logs for each thread).

**Comment by hainm** [comment]
> or just cpptraj.OMP

with serial version too
I tried

`~/amber_git/amber/bin/cpptraj  -i tz2.in` and got `Segmentation fault` in the end of the log.

From my experience discussing with @drroe, I find that notify him first and do digging later :D since he always knows what really going on. ha ha.

**Comment by hainm** [comment]
> What I do when I get a segfault is to compile with the "-g" flag turned on (which adds debug symbols), then issue the command ulimit -c unlimited. This means that a core dump will be written no matter how large it is (I think default size is 0, so that a core.pid file is never written). You can then load this into gdb and type bt to get a stack trace (which will tell you the exact source line where the segfault occurred with the whole call stack).

> If you don't want to deal with the core file (which is really the best way to debug parallel segfaults), you can run cpptraj in gdb and just use run <arguments>. The debugger will stop at the segfault, and you can type bt to get the stack trace.

> This removes the guessing, usually, and usually lets you debug issues yourself. If you still can't figure it out (as has happened to me a number of times), I post with the full stack trace, which helps whoever does figure it out quite a bit.

great. thanks. Those things are quite new to me. Will try to learn gradually.

**Comment by hainm** [comment]
oh, just tried 

gdb with run tips from your post, sounds cool to me.

```
Program received signal SIGSEGV, Segmentation fault.
0x000000000048e2a4 in Action_NAstruct::Setup(ActionSetup&) ()
```

**Comment by swails** [comment]
> From my experience discussing with @drroe, I find that notify him first and do digging later :D since he always knows what really going on. ha ha.

In this case, that's probably the case.  But what about when you're debugging an obscure sander or pmemd segfault (gasp! with MPI, even!) and the original author of the code is... even if you know *who*, long since gone?  You can always pass it off, but debugging is a useful skill (and in Amber, right now there are a small number of people that will actually pick up difficult Heisenbugs or obscure segfaults, and there is no guarantee we will stick around forever).

The tips I gave don't take much time once you're experienced with it and are well worth doing IMO.  For instance, it's why I found #153 so easy to fix (yes, the action was originally my code, but the part that was segfaulting has changed a bit as a result of Dan's retooling).

**Comment by hainm** [comment]
> In this case, that's probably the case. But what about when you're debugging an obscure sander or pmemd segfault (gasp! with MPI, even!) and the original author of the code is... even if you know who, long since gone? You can always pass it off, but debugging is a useful skill (and in Amber, right now there are a small number of people that will actually pick up difficult Heisenbugs or obscure segfaults, and there is no guarantee we will stick around forever).

got it. thanks.

**Comment by swails** [comment]
> gdb with run tips from your post, sounds cool to me.

If you don't compile with `-g`, you don't get line numbers in your stack traces -- you just get memory addresses.

And I'm not fluent in hexadecimal garbage :).  I know how to tell whether or not variables are close in memory (see how close the hexadecimal numbers are :)) which is sometimes useful, but that's like reading at a toddler level.  Line numbers in the source code is much easier for me to read ;).

**Comment by hainm** [comment]
OK, so I guess workflow to debug segmentation fault is 

- compile program with `-g` flag 

-  run gdb: `gdb ../../cpptraj/bin/cpptraj.OMP`

- type `run`

- enter some commands (cpptraj's ones as example)

- get segfault

- run `bt` command

**Comment by swails** [comment]
That would do it.  If you have a cpptraj script, you could do

```
run -i cpptraj.input [-p prmtop]
```

where it takes cpptraj command-line.

**Comment by hainm** [comment]
@swails 

just recompiled cpptraj with `-g` flag, re-debug and I can identify the line number.  thanks.

```
Program received signal SIGSEGV, Segmentation fault.
Action_NAstruct::Setup (this=0x984d20, setup=...) at Action_NAstruct.cpp:1210
1210      std::vector<int> Visited( setup.Top().Res(Bases_.back().ResNum()).LastAtom(), 0 );
```

> run -i cpptraj.input [-p prmtop]

:+1:

**Comment by swails** [comment]
:)

**Comment by hainm** [comment]
just added `CpptrajDevlopmentGuide` to try to add this debug info in future.

**Comment by swails** [comment]
You spelled development wrong ;)

**Comment by hainm** [comment]
> You spelled development wrong ;)

ha ha, @drroe  is the one to be blamed. Strictly copied and pasted from his filename. :))

**Comment by drroe** [comment]
Will be fixed soon.

**Comment by swails** [comment]
Magic Dan

**Comment by hainm** [comment]
> Magic Dan

told you. :D

**Comment by drroe** [comment]
Should be fixed by #169

**Comment by hainm** [comment]
confirmed.
